src/NuGet.Services.BasicSearch/ApplicationInsightsMiddleware.cs

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Threading.Tasks;
+using Microsoft.ApplicationInsights;
+using Microsoft.ApplicationInsights.DataContracts;
+using Microsoft.ApplicationInsights.Extensibility;
+using Microsoft.Owin;
+
+namespace NuGet.Services.BasicSearch
+{
+    public class ApplicationInsightsMiddleware
+        : OwinMiddleware
+    {
+        public const string OwinRequestIdKey = "owin.RequestId";
+
+        private readonly TelemetryClient _telemetryClient;
+
+        public ApplicationInsightsMiddleware(OwinMiddleware next)
+            : this(next, null)
+        {
+        }
+
+        public ApplicationInsightsMiddleware(OwinMiddleware next, TelemetryConfiguration telemetryConfiguration) 
+            : base(next)
+        {
+            if (next == null)
+            {
+                throw new ArgumentNullException(nameof(next));
+            }
+
+            _telemetryClient = telemetryConfiguration == null 
+                ? new TelemetryClient()
+                : new TelemetryClient(telemetryConfiguration);
+        }
+
+        public override async Task Invoke(IOwinContext context)
+        {
+            var requestId = context.Get<string>(OwinRequestIdKey);
+
+            var requestMethod = context.Request.Method;
+            var requestPath = context.Request.Path.ToString();
+            var requestUri = context.Request.Uri;
+
+            var requestStartDate = DateTimeOffset.Now;
+            var stopWatch = new Stopwatch();
+            stopWatch.Start();
+
+            var requestFailed = false;
+
+            try
+            {
+                OwinRequestIdContext.Set(requestId);
+
+                await Next.Invoke(context);
+            }
+            catch (Exception)
+            {
+                requestFailed = true;
+                throw;
+            }
+            finally
+            {
+                stopWatch.Stop();
+
+                TrackRequest(requestId, requestMethod, requestPath, requestUri, 
+                    context.Response.StatusCode, requestFailed, requestStartDate, stopWatch.Elapsed);
+
+                OwinRequestIdContext.Clear();
+            }
+        }
+
+        private void TrackRequest(string requestId, string requestMethod, string path, Uri uri, int responseCode,
+            bool requestFailed, DateTimeOffset requestStartDate, TimeSpan duration)
+        {
+            var name = $"{requestMethod} {path}";
+
+            var telemetry = new RequestTelemetry
+            {
+                Id = requestId,
+                Name = name,
+                StartTime = requestStartDate,
+                Duration = duration,
+                ResponseCode = responseCode.ToString(),
+                Success = responseCode < 400 || !requestFailed,
```

Comment:
Switched to >= 200 && <= 299 (as per https://msdn.microsoft.com/en-us/library/windows/apps/windows.web.http.httpresponsemessage.issuccessstatuscode)